src/postgraphile/postgraphile.ts

```diff
@@ -100,22 +89,52 @@ export function getPostgraphileSchemaBuilder<
 
   async function createGqlSchema(): Promise<GraphQLSchema> {
     let attempts = 0;
+    let isShuttingDown = false;
+    shutdownActions.add(async () => {
+      isShuttingDown = true;
+    });
     // eslint-disable-next-line no-constant-condition
     while (true) {
+      if (isShuttingDown) {
+        throw new Error('Postgraphile is shutting down');
+      }
       try {
         if (options.watchPg) {
-          await watchPostGraphileSchema(pgPool, pgSchemas, options, newSchema => {
-            gqlSchema = newSchema;
-            _emitter.emit('schemas:changed');
-            exportGqlSchema(gqlSchema);
+          const releaseWatchFnPromise = watchPostGraphileSchema(
+            pgPool,
+            pgSchemas,
+            options,
+            newSchema => {
+              gqlSchema = newSchema;
+              _emitter.emit('schemas:changed');
+              exportGqlSchema(gqlSchema);
+            },
+          );
+          // We must register the shutdown action immediately to avoid a race condition.
+          shutdownActions.add(async () => {
```

Comment:
This registers a shutdown action during the loop but never releases it; this captures the closure (and its variables) forever and could result in memory exhaustion. I've refactored the code to only register the shutdown actions outside the loop.